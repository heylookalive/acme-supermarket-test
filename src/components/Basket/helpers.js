export const basketToProductsAndQuantities = (items) => items.reduce((acc, product) => {
  if (typeof acc[product.code] === 'undefined') {
    acc[product.code] = {
      name: product.name,
      price: product.price,
      quantity: 1,
    };
  } else {
    acc[product.code].quantity += 1;
  }
  return acc;
}, {});

export const getBasketItemDiscounts = (basketItems, pricingRules) => {
  const lineItems = [];
  let totalDiscount = 0;
  // Create a simple array of which products we have and the quantity of each.
  const productsAndQuantities = basketToProductsAndQuantities(basketItems);

  pricingRules.forEach(rule => {
    if (productsAndQuantities[rule.appliesTo]) {
      const applicableProduct = productsAndQuantities[rule.appliesTo];
      // Buy one get one free will only apply when there's more than 1!
      if (rule.type === 'BOGOF' && applicableProduct.quantity > 1) {
        // Hurt my head more than it should, given a number, find the closest
        // even number rounding down to then discount.
        const closestEvenQuantity =
          2 * Math.floor(applicableProduct.quantity / 2);
        const lineItemDiscount =
          (closestEvenQuantity / 2) * applicableProduct.price;
        lineItems.push({
          id: rule.id,
          name: rule.name,
          productName: applicableProduct.name,
          discount: lineItemDiscount,
        });
        totalDiscount += lineItemDiscount;
      } else if (
        rule.type === 'PRICE_DISCOUNT' &&
        applicableProduct.quantity >= rule.minimumProducts
      ) {
        // Price discounts as I read the instructions will apply to all units
        // of that product, when a minimum number of that product are added
        // to the basket.
        const lineItemDiscount =
          (applicableProduct.price - rule.discountUnitTo) *
          applicableProduct.quantity;
        lineItems.push({
          id: rule.id,
          name: rule.name,
          productName: applicableProduct.name,
          discount: lineItemDiscount,
        });
        totalDiscount += lineItemDiscount;
      }
    }
  });
  return {
    lineItems,
    totalDiscount,
  };
};
