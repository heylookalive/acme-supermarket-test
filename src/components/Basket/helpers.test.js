import { basketToProductsAndQuantities, getBasketItemDiscounts } from './helpers';

describe('basketToProductsAndQuantities helper', () => {
  it('returns a correct array of products and quantities', () => {
    const basketItems = [{
      code: 'TST123',
      name: 'Test 123',
      price: 1
    },
    {
      code: 'TST123',
      name: 'Test 123',
      price: 1
    },
    {
      code: 'TST234',
      name: 'Test 234',
      price: 1
    }];

    expect(basketToProductsAndQuantities(basketItems)).toMatchSnapshot();
  });
});

describe('getBasketItemDiscounts helper', () => {
  // Was on the fence between importing the data from the file which is static
  // knowing it won't change in this exercise, but want it static and in the
  // test as static.
  const pricingRules = [
    {
      "id": "pr1",
      "name": "CEO Loves Fruit tea BOGOF",
      "type": "BOGOF",
      "appliesTo": "FR1"
    },
    {
      "id": "pr2",
      "name": "Strawberry bulk discount",
      "type": "PRICE_DISCOUNT",
      "appliesTo": "SR1",
      "minimumProducts": 3,
      "discountUnitTo": 4.5
    }
  ];
  const fruitTeaProduct = {
    code: "FR1",
    name: "Fruit tea",
    price: 3.11
  };
  const strawberryProduct =   {
    code: "SR1",
    name: "Strawberries",
    price: 5.00
  };
  const coffeeProduct = {
    code: "CF1",
    name: "Coffee",
    price: 11.23
  }

  it('returns the correct values given fruit tea bogof and basket with 2', () => {
    const discounts = getBasketItemDiscounts(Array(2).fill(fruitTeaProduct), pricingRules);
    expect(discounts.lineItems.length).toEqual(1);
    expect(discounts.totalDiscount).toEqual(3.11);
  });

  it('returns the correct values given fruit tea bogof and basket with 3', () => {
    const discounts = getBasketItemDiscounts(Array(3).fill(fruitTeaProduct), pricingRules);
    expect(discounts.lineItems.length).toEqual(1);
    expect(discounts.totalDiscount).toEqual(3.11);
  });

  it('returns the correct values given fruit tea bogof and basket with 6', () => {
    const discounts = getBasketItemDiscounts(Array(6).fill(fruitTeaProduct), pricingRules);
    expect(discounts.lineItems.length).toEqual(1);
    expect(discounts.totalDiscount).toEqual(9.33);
  });

  it('returns the correct values given 2x fruit tea, 1 coffee and 1 strawberry', () => {
    const discounts = getBasketItemDiscounts([fruitTeaProduct, strawberryProduct, fruitTeaProduct, coffeeProduct], pricingRules);
    expect(discounts.lineItems.length).toEqual(1);
    expect(discounts.totalDiscount).toEqual(3.11);
  });

  it('returns the correct values given 3x strawberry and 1 fruit tea', () => {
    const discounts = getBasketItemDiscounts([strawberryProduct, fruitTeaProduct, strawberryProduct, strawberryProduct], pricingRules);
    expect(discounts.lineItems.length).toEqual(1);
    expect(discounts.totalDiscount).toEqual(1.50);
  });

  it('returns the correct values given 3x strawberry and 4x fruit tea', () => {
    const discounts = getBasketItemDiscounts([strawberryProduct, fruitTeaProduct, strawberryProduct, strawberryProduct, fruitTeaProduct, fruitTeaProduct, fruitTeaProduct], pricingRules);
    expect(discounts.lineItems.length).toEqual(2);
    expect(discounts.totalDiscount).toEqual(7.72);
  });
});
