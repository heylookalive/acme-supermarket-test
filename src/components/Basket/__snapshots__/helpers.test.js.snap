// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basketToProductsAndQuantities helper returns a correct array of products and quantities 1`] = `
Object {
  "TST123": Object {
    "name": "Test 123",
    "price": 1,
    "quantity": 2,
  },
  "TST234": Object {
    "name": "Test 234",
    "price": 1,
    "quantity": 1,
  },
}
`;
